### DISCUSSION TITLE
Action runs when PR is created, but not when PR is merged

### DISCUSSION THREAD

**Discussion by jrobison-sb** [community/community#25073] [Actions]
<p>Hi,</p>
<p>This is the same issue as described in support ticket 358105, but it’s been 11 days and I haven’t had any reply on that ticket. I have links to relevant PR’s and stuff in that ticket if any staff would find those useful.</p>
<p>I have a GitHub action which runs when a PR is created, and runs when I add reviewers to the PR, all of which is expected. But it does not run when the PR gets merged. And given that this is the jessfraz branch cleanup action, the only relevant event for this action is when a PR gets merged. When a PR gets merged, there’s no logs or any indication in the ‘Actions’ UI that the action runs, though I do see logs for it when a PR is created and reviewers are added.</p>
<p>Note that this worked great on the old HCL system, then it stopped running entirely for all events a few weeks ago. I have since migrated it to YAML and it now runs when a PR is created, but not when the PR is merged. In case of any of that is relevant.</p>
<p>Anybody know why this isn’t running when a PR gets merged? Thanks.</p>
<p>Here’s my YAML:</p>
<pre><code>on: pull_request
name: on pull request merge, delete the branch
jobs:
  branchCleanup:
    name: branch cleanup
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master
    - name: branch cleanup
      uses: jessfraz/branch-cleanup-action@master
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        NO_BRANCH_DELETED_EXIT_CODE: 0
</code></pre>

**Comment by chrispat** [comment]
<p>By default a workflow is run when a PR is opened, synchronized or reopened to better optimize for CI workflows.  If you want to run when the PR is merged you can specify that as follows.</p>
<pre><code>on:
  pull_request:
    types: [merged]
</code></pre>
<p>Since events are not really individual events but are a set of events that could occur on a givne resources (at lest mostly).  Using the types: field you can filter against the action element in the payload.</p>

**Comment by jrobison-sb** [comment]
<p>Thanks for your reply. I adjusted my <code>.github/workflows/pull_request.yml</code> file as you suggested, but it still didn’t run upon merging a PR. Any further suggestion or insight?</p>
<pre><code>on:
  pull_request:
    types: [merged]
name: on pull request merge, delete the branch
jobs:
  branchCleanup:
    name: branch cleanup
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master
    - name: branch cleanup
      uses: jessfraz/branch-cleanup-action@master
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        NO_BRANCH_DELETED_EXIT_CODE: 0
</code></pre>

**Comment by talldan** [comment]
<p>Should that be ‘closed’? e.g.:</p>
<pre><code>on:
  pull_request:
    types: [closed]
</code></pre>
<p>According to the docs there isn’t a ‘merged’ type for a pull_request action:<br>
<a href="https://help.github.com/en/articles/events-that-trigger-workflows" rel="nofollow noopener">https://help.github.com/en/articles/events-that-trigger-workflows</a></p>

**Comment by cdb** [comment]
<p><a class="mention" href="/jrobison-sb">@jrobison-sb</a>  - <a class="mention" href="/talldan">@talldan</a> was exactly right, it should be <code>closed</code>, not <code>merged</code>.</p>
<p>However, even that’s not going to work right now as you’ve uncovered a bug in the system where actions aren’t triggering on a closed PR even when they should be (using this exact setup). I’ve been able to recreate it and I have a fix in the works. I’ll report back here when I have something deployed and working!</p>
<p>(btw, I’m an engineer on the Actions team, not sure if my profile says that)</p>

**Comment by talldan** [comment]
<p>Thanks <a class="mention" href="/cdb">@cdb</a>, that fix and the <code>types</code> option would make us very happy :slight_smile:</p>

**Comment by cdb** [comment, marked as answer]
<p>Awesome <a class="mention" href="/talldan">@talldan</a>, good to know! Fixing things makes me happy too :) </p>
<p><a class="mention" href="/jrobison-sb">@jrobison-sb</a>  The immediate bug is deployed and working now. It uncovered another issue which I’m going to work on, but the workaround is to simply remove the following line completely:</p>
<pre><code>- uses: actions/checkout@master
</code></pre>
<p>It’s not working right now in this use case (something we have to fix on our side), but for this specific action you don’t actually have to checkout your code so removing it should get things running I believe!</p>

**Comment by jrobison-sb** [comment]
<p>@ cbd Thanks for identifying this, and thanks for the super quick turn around time on the fix.</p>
<p>It’s working again as expected. </p>

**Comment by ghost** [comment]
<blockquote>
<p><a class="mention" href="/cdb">@cdb</a> wrote:</p>
</blockquote>
<blockquote>
<p><a class="mention" href="/jrobison-sb">@jrobison-sb</a>  - <a class="mention" href="/talldan">@talldan</a> was exactly right, it should be <code>closed</code>, not <code>merged</code>.</p>
</blockquote>
<p>Would it be possible to create a <code>merged</code> type? I’m building a pipeline that automatically creates a release when a release/* branch is merged into master. But closed (canceled) pull requests will also trigger this build.</p>

**Comment by sanohin** [comment]
<p>And what about checkout? Is there any workaround if I need to get the code from the merged branch for the pipeline? </p>

**Comment by vvakame** [comment]
<p>are there any updates?</p>
<p>I’m just trying excute workflow when PR merged.</p>
<p>But it doesn’t kicked from platform…</p>
<p><a href="https://github.com/vvakame/til/blob/master/.github/workflows/pr-to-blog.yml" rel="nofollow noopener">https://github.com/vvakame/til/blob/master/.github/workflows/pr-to-blog.yml</a></p>

**Comment by caraya** [comment]
<p>If I’m reading the documentation correctly, closed will trigger on any reason for closing a PR, so my action could be triggered when a PR is closed without merging which is not what I need. </p>
<p>Am I missing something? Is there a way to trigger the action on a closed PR that has been merged?</p>
